tests/fourier/test_reconstruct.py

```diff
@@ -0,0 +1,755 @@
+
+        rec = _reconstruct_gen(Fun, spectrum)
+        assert spy.call_count == len([f for f in spectrum if f > 0.0]) * 2 + 1
+        assert fun_close(circuit, rec)
+
+        # Repeat, using precomputed f0
+        f0 = circuit(0.0)
+        Fun = Lambda(circuit)
+        spy = mocker.spy(Fun, "fun")
+        rec = _reconstruct_gen(Fun, spectrum, f0=f0)
+        assert spy.call_count == len([f for f in spectrum if f > 0.0]) * 2
+        assert fun_close(circuit, rec)
+
+
+all_ids = [
+    None,
+    {"x": [0, 1], "y": {1, 5}},
+    {"z": (0, 9)},
+    ["z", "y", "x", "x"],
+    {"x", "z", "y"},
+    ("y",),
+    "x",
+]
+
+all_spectra = [
+    None,
+    {
+        "x": {0: [0.0], 1: [4.2, 0.0, 0.2]},
+        "y": {3: [0.3, 0.0, 0.2], 1: [0.0, 1.1, 5.2], 5: [0.0, 1.2]},
+        "z": {i: [0.0, i * 8.7] for i in range(20)},
+    },
+]
+
+all_shifts = [
+    None,
+    {
+        "x": {0: [-1.3], 1: [1.0, -0.4, 4.2, 2.3, -1.5]},
+        "y": {
+            3: [0.3 * i + 0.05 for i in range(-2, 3)],
+            1: [-1, -0.5, -0.1, 0.1, 0.9],
+            5: [-1, -0.5, -0.2],
+        },
+        "z": {i: [-np.pi / 2, 0.0, np.pi / 2] for i in range(20)},
+    },
+]
+
+all_nums_frequency = [
+    None,
+    {
+        "x": {0: 1, 1: 4},
+        "y": {3: 1, 1: 1, 5: 9},
+        "z": {i: 2 * i for i in range(20)},
+    },
+]
+
+
+class TestPrepareJobs:
+    """Tests the subroutine that determines the 1D reconstruction
+    jobs to be carried out for a call to ``reconstruct`` ."""
+
+    @pytest.mark.parametrize("ids", all_ids)
+    @pytest.mark.parametrize("spectra", all_spectra)
+    @pytest.mark.parametrize("shifts", all_shifts)
+    @pytest.mark.parametrize("nums_frequency", all_nums_frequency)
+    def test_prepjobs(self, ids, spectra, shifts, nums_frequency, tol):
+        """Test ``_prepare_jobs`` with a large variety of test cases (cheap)."""
+        if nums_frequency is None and spectra is None:
+            with pytest.raises(ValueError, match="Either nums_frequency or spectra"):
+                _prepare_jobs(ids, nums_frequency, spectra, shifts, atol=tol)
+            return
+
+        ids_, recon_fn, jobs, need_f0 = _prepare_jobs(
+            ids,
+            nums_frequency,
+            spectra,
+            shifts,
+            atol=tol,
+        )
+
+        # Check ids
+        if ids is None:
+            # Test automatic generation of ids
+            if nums_frequency is None:
+                assert ids_.keys() == spectra.keys()
+                for _id, _ids in ids_.items():
+                    assert _ids == spectra[_id].keys()
+            else:
+                assert ids_.keys() == nums_frequency.keys()
+                for _id, _ids in ids_.items():
+                    assert _ids == nums_frequency[_id].keys()
+
+        else:
+            assert all(id_ in ids for id_ in ids_.keys())
+            if type(ids) == dict:
+                assert all(ids_[id_] == ids[id_] for id_ in ids_.keys())
```

Comment:
@dwierichs this definitely makes me think this test should be split into multiple tests, one per branch!

Reply:
I was thinking the same, but there are very many cases to be checked, so I went for this mess.. I'll try to disentangle it :)